src/librustc_trans/trans/mir/block.rs

```diff
@@ -46,8 +47,25 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
                 build::CondBr(bcx, cond.immediate(), lltrue, llfalse, DebugLoc::None);
             }
 
-            mir::Terminator::Switch { .. } => {
-                unimplemented!()
+            mir::Terminator::Switch { ref discr, ref adt_def, ref targets } => {
+                let adt_ty = bcx.tcx().lookup_item_type(adt_def.did).ty;
+                let represented_ty = adt::represent_type(bcx.ccx(), adt_ty).clone();
+
+                let discr_lvalue = self.trans_lvalue(bcx, discr);
+                let discr = adt::trans_get_discr(bcx, &*represented_ty, discr_lvalue.llval, None);
+
+                // The else branch of the Switch can't be hit, so branch to an unreachable
+                // instruction so LLVM knows that
+                let unreachable_blk = bcx.fcx.new_temp_block("enum-variant-unreachable");
+                build::Unreachable(unreachable_blk);
```

Comment:
As @nagisa says, it might be nice to have just one such block (created lazilly), we could store it in the "MIR trans" state.